src/endpoint/s3/ops/s3_post_object_select.js

```diff
@@ -0,0 +1,219 @@
+ */
+// eslint-disable-next-line max-statements
+async function post_object_select(req, res) {
+
+    //if (('S3Select' in nb_native()) === false) {
+    if (nb_native().S3Select === undefined) {
+        dbg.error("Request for object select but compiled without BUILD_S3SELECT=1");
+        res.statusCode = 501;
+        res.end();
+        return;
+    }
+
+    req.object_sdk.setup_abort_controller(req, res);
+    const agent_header = req.headers['user-agent'];
+    const noobaa_trigger_agent = agent_header && agent_header.includes('exec-env/NOOBAA_FUNCTION');
+    const encryption = s3_utils.parse_encryption(req);
+
+    const md_params = {
+        bucket: req.params.bucket,
+        key: req.params.key,
+        version_id: req.query.versionId,
+        encryption,
+    };
+    if (req.query.get_from_cache !== undefined) {
+        md_params.get_from_cache = true;
+    }
+    const object_md = await req.object_sdk.read_object_md(md_params);
+
+    const params = {
+        object_md,
+        obj_id: object_md.obj_id,
+        bucket: req.params.bucket,
+        key: req.params.key,
+        content_type: object_md.content_type,
+        noobaa_trigger_agent,
+        encryption,
+    };
+    if (md_params.get_from_cache) {
+        params.get_from_cache = true;
+    }
+
+    const read_stream = await req.object_sdk.read_object_stream(params, res);
+    if (read_stream) {
+        // if read_stream supports closing, then we handle abort cases such as http disconnection
+        // by calling the close method to stop it from buffering more data which will go to waste.
+        if (read_stream.close) {
+            req.object_sdk.add_abort_handler(() => read_stream.close());
+        }
+        read_stream.on('error', err => {
+            dbg.log0('read stream error:', err, req.path);
+            res.destroy(err);
+        });
+        const http_req_select_params = req.body.SelectObjectContentRequest;
+        const input_serialization = http_req_select_params.InputSerialization[0];
+        let input_format = null;
+        if ('CSV' in input_serialization) {
+            input_format = 'CSV';
+        } else if ('JSON' in input_serialization) {
+            input_format = 'JSON';
+        } else {
+            throw new S3Error(S3Error.InvalidArgument);
+        }
+        const select_args = {
+            query: http_req_select_params.Expression[0],
+            input_format: input_format,
+            input_serialization_format: http_req_select_params.InputSerialization[0][input_format][0],
+            http_res: res
+        };
+        const s3select = new S3SelectStream(select_args);
+        /*s3select.on('end', () => {
+            res.write(end_message);
+            res.end();
+        });*/
+        dbg.log3("select_args = ", select_args);
+        read_stream.pipe(new EchoDebugStream({header: "before = "})).pipe(s3select).pipe(new EchoDebugStream({header: "after = "})).pipe(res);
```

Comment:
calling .pipe.pipe.pipe is not handling all the error propagation and destroyed cases well enough. for more context see this test code - [test_pipeline_stuck_on_destroyed_stream.js](https://github.com/noobaa/noobaa-core/blob/master/src/test/unrelated/test_pipeline_stuck_on_destroyed_stream.js).

this is why we have our pipeline function in [stream_utils.js](https://github.com/noobaa/noobaa-core/blob/579969198fff7c53a68a5372080040ef8b8a5ff4/src/util/stream_utils.js#L32) that wraps util.pipeline and adds some more checks.